File: 3514-number-of-unique-xor-triplets-ii/3514-number-of-unique-xor-triplets-ii.go
func uniqueXorTriplets(nums []int) int {
    // xor operation result can't exceed the longest bit
    // 1500 = 10111011100 -> max: 11111111111 (11) = 2047
    const MAX = 2047

    a, b, c := make([]bool, MAX+1), make([]bool, MAX+1), make([]bool, MAX+1)

    for _, n := range nums {
        a[n] = true
    }

    for i := 0; i < MAX; i++ {
        if !a[i] {
            continue
        }
        for _, n := range nums {
            b[i^n] = true
        }
    }

    for i := 0; i < MAX; i++ {
        if !b[i] {
            continue
        }
        for _, n := range nums {
            c[i^n] = true
        }
    }

    ans := 0
    for _, ok := range c {
        if ok {
            ans++
        }
    }

    return ans
}

// array, brute force + optimization
// time: O(MAX * n)
// space: O(MAX)

File: 3514-number-of-unique-xor-triplets-ii/3514-number-of-unique-xor-triplets-ii.py
class Solution:
    def uniqueXorTriplets(self, nums: List[int]) -> int:
        MAX = 2047

        a, b, c = [False] * (MAX + 1), [False] * (MAX + 1), [False] * (MAX + 1)

        for n in nums:
            a[n] = True
        
        for i in range(MAX):
            if not a[i]:
                continue

            for n in nums:
                b[i ^ n] = True
        
        for i in range(MAX):
            if not b[i]:
                continue
            
            for n in nums:
                c[i ^ n] = True
        
        ans = 0
        for ok in c:
            if ok:
                ans += 1
        
        return ans

# array, brute force + optimization
# time: O(MAX * n)
# space: O(MAX)